etc/scripts/timetable/old-stack-import/augment-tree-with-user-shibboleth-ids.js

```diff
@@ -0,0 +1,238 @@
+var argv = yargs
+    .usage('Augment the event organisers with their shibboleth ids.\nUsage: $0')
+    .example('$0 --input tree.json --output tree-with-shibboleth-ids.json --users users.csv')
+
+    .demand('i')
+    .alias('i', 'input')
+    .describe('i', 'The path where the JSON tree can be read')
+
+    .demand('o')
+    .alias('o', 'output')
+    .describe('o', 'The path where the augmented JSON tree should be written to')
+
+    .demand('u')
+    .alias('u', 'users')
+    .describe('u', 'The CSV file containing all the users. The expected columns are: dn, uid, cn, displayName, mail')
+    .argv;
+
+// Keep a cache of matched names to user records
+var matched = {};
+
+// The number of events that have been visited by the script
+var visittedEvents = 0;
+
+/**
+ * Visit a node in the tree. If the node is an event, the organisers will be looked up in the set
+ * of users. In case there's a match, the organiser will be replaced by an object referring to the
+ * user. In case the node is not an event, its children will be visited.
+ *
+ * As this could potentially take a very long time, the entire tree will be flushed to disk every while
+ *
+ * @param  {Object}         node        The node to visit
+ * @param  {Object[]}       users       The set of users to match the organisers with
+ * @param  {Object}         tree        The full tree who will be modified in-place
+ * @api private
+ */
+var visitNode = function(node, users, tree) {
+    if (node.type === 'event') {
+        if (!node.peopleSeen) {
+            _.each(node.people, function(organiser, index) {
+                var match = findMatch(organiser, users);
+                if (match) {
+                    node.people[index] = {
+                        'shibbolethId': match.uid + '@cam.ac.uk',
+                        'displayName': match.displayName,
+                        'original': organiser
+                    };
+                }
+            });
+            node.peopleSeen = true;
+        }
+        visittedEvents++;
+        if (visittedEvents % 25 === 0) {
+            console.log('Handled %d events', visittedEvents);
+
+            // Periodically write the tree to disk
+            writeTree(tree);
+        }
+    } else {
+        _.each(node.nodes, function(child) {
+            visitNode(child, users, tree);
+        });
+    }
+};
+
+/**
+ * Write the tree to disk
+ *
+ * @param  {Object}     tree    The tree that should be flushed to disk
+ * @api private
+ */
+var writeTree = function(tree) {
+    fs.writeFileSync(argv.output, JSON.stringify(tree, null, 4));
+};
+
+/**
+ * Given a name, find the best matching user
+ *
+ * @param  {String}     name    The name to find a user for
+ * @param  {Object}     users   The set of users
+ * @return {Object}             The best matching user, or null if no user could be matched
+ * @api private
+ */
+var findMatch = function(name, users) {
+    var bestMatch = [];
+    var max = 100000;
+
+    if (matched[name]) {
```

Comment:
I see, we only populate ones that have a unique best match. Can you maybe document this process a bit?